packages/components/hooks/use-scroll-position.ts

```diff
@@ -0,0 +1,29 @@
+import { useRef, useState } from 'react'
+
+import { useIsomorphicLayoutEffect } from './use-isomorphic-layout-effect'
+
+interface ScrollPosition {
+  ref: React.RefObject<HTMLDivElement>
+  scrollPos: number
+}
+
+const useScrollPosition = (): ScrollPosition => {
+  const ref = useRef<HTMLDivElement>(null)
+  const [scrollPos, setScrollPos] = useState(0)
+
+  useIsomorphicLayoutEffect(() => {
+    const newRef = ref.current
+    const handleScroll = () => {
+      setScrollPos(newRef!.scrollTop)
+    }
+    newRef!.addEventListener('scroll', handleScroll)
```

Comment:
Scroll events are firing at a very high rate. Therefore it shouldn't trigger anything computationally expensive. Currently, it is setting the new state and re-rendering the whole application for every event.

Few recommendations: 
- always throttle the number of times the event handler is called
- use passive flag for better scroll performance, more on that [here](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#improving_scrolling_performance_with_passive_listeners)
- set state only based on your condition 
- use intersection observer to let browser call your handler

Reply:
Can you show an example an example of the last point with this kind of use? I've already use intersection observers for infinite scroll but i would love to have a better insight of this approach. 😄